### PR TITLE
ci/installer-5327 : aws-ec2 - change instance to m6i

### PR DESCRIPTION
**Pull request by mtulio**
fixes for installer PR https://github.com/openshift/installer/pull/5327

The idea is to choose the preferred instance type (in general newer gen) when it's available on the equal or higher number of zones on a given region, otherwise, the backup will be chosen.

/cc  @staebler

### CODE REVIEW

File: ci-operator/step-registry/ipi/conf/aws/ipi-conf-aws-commands.sh [hunk 1]
```diff
@@ -13,31 +13,82 @@ CONFIG="${SHARED_DIR}/install-config.yaml"
...
+  local zones_cnt_pref
+  local zones_cnt_backup
+
+  type_preferred="${1}.$(get_master_size)"
+  type_backup="${2}.$(get_master_size)"
+
+  zones_cnt_pref=$(aws ec2 describe-instance-type-offerings --region "${REGION}" --location-type availability-zone --filters Name=instance-type,Values="${type_preferred}" --query 'InstanceTypeOfferings[].Location' --output json | jq -r '. | length')
+  zones_cnt_backup=$(aws ec2 describe-instance-type-offerings --region "${REGION}" --location-type availability-zone --filters Name=instance-type,Values="${type_backup}" --query 'InstanceTypeOfferings[].Location' --output json | jq -r '. | length')
+
+  if [[ ${zones_cnt_pref} -ge ${zones_cnt_backup} ]]; then
```
Comment (staebler) [inline ci-operator/step-registry/ipi/conf/aws/ipi-conf-aws-commands.sh:56]: In practice this probably ends up the same, but I would have expected that the preferred family be used as long as there was at least one zone (or perhaps 2 zones) offering the preferred family.

File: ci-operator/step-registry/ipi/conf/aws/ipi-conf-aws-commands.sh [hunk 2]
```diff
@@ -13,31 +13,82 @@ CONFIG="${SHARED_DIR}/install-config.yaml"
...
+    master_size=4xlarge
+  elif [[ "${SIZE_VARIANT}" == "compact" ]]; then
+    master_size=2xlarge
+  fi
+  echo ${master_size}
+}
+
+function set_master_type {
+  local size
+  master_type=null
```
Comment (staebler) [inline ci-operator/step-registry/ipi/conf/aws/ipi-conf-aws-commands.sh:35]: Why is this necessary?

File: ci-operator/step-registry/ipi/conf/aws/ipi-conf-aws-commands.sh [hunk 3]
```diff
@@ -13,31 +13,82 @@ CONFIG="${SHARED_DIR}/install-config.yaml"
...
+  local type_preferred
+  local type_backup
+  local zones_cnt_pref
+  local zones_cnt_backup
+
+  type_preferred="${1}.$(get_master_size)"
+  type_backup="${2}.$(get_master_size)"
+
+  zones_cnt_pref=$(aws ec2 describe-instance-type-offerings --region "${REGION}" --location-type availability-zone --filters Name=instance-type,Values="${type_preferred}" --query 'InstanceTypeOfferings[].Location' --output json | jq -r '. | length')
+  zones_cnt_backup=$(aws ec2 describe-instance-type-offerings --region "${REGION}" --location-type availability-zone --filters Name=instance-type,Values="${type_backup}" --query 'InstanceTypeOfferings[].Location' --output json | jq -r '. | length')
```
Comment (staebler) [inline ci-operator/step-registry/ipi/conf/aws/ipi-conf-aws-commands.sh:54]: This does not take into account whether AZs are actually available or whether the AZs were specified already in the install-config.yaml.

File: ci-operator/step-registry/ipi/conf/aws/ipi-conf-aws-commands.sh (hunk 1 above)
Comment (mtulio) [reply ci-operator/step-registry/ipi/conf/aws/ipi-conf-aws-commands.sh:56]: iiuc, I am not sure if it will ends up the same, it will help to select the correct instance on the region which does not support the preferred one (see the [original PR](https://github.com/openshift/installer/pull/5327) for the table of availability):

output of `$ cat install-config.yaml |yq -r '. | (.platform.aws, .controlPlane.platform.aws)'`,

for `LEASED_RESOURCE=sa-east-1`:
``` json
{
  "region": "sa-east-1",
  "userTags": {
    "expirationDate": "2021-12-01T22:34+00:00"
  }
}
{
  "zones": [],
  "type": "m5.2xlarge"
}

```

for `LEASED_RESOURCE=us-east-1`:
```json
{
  "region": "us-east-1",
  "userTags": {
    "expirationDate": "2021-12-01T22:35+00:00"
  }
}
{
  "zones": [],
  "type": "m6i.2xlarge"
}
```

Comment (mtulio) [reply ci-operator/step-registry/ipi/conf/aws/ipi-conf-aws-commands.sh:56]: > but I would have expected that the preferred family be used as long as there was at least one zone (or perhaps 2 zones)

if it's not a problem, we could change to something like:

``` shell
if [[ ${zones_cnt_pref} -ge 1 ]]; then
```

wdyt?

File: ci-operator/step-registry/ipi/conf/aws/ipi-conf-aws-commands.sh (hunk 2 above)
Comment (mtulio) [reply ci-operator/step-registry/ipi/conf/aws/ipi-conf-aws-commands.sh:35]: it's not. I just removed it, and replace the `${master_type}` to `${MASTER_TYPE}` used in this script. 👍🏼

File: ci-operator/step-registry/ipi/conf/aws/ipi-conf-aws-commands.sh (hunk 3 above)
Comment (mtulio) [reply ci-operator/step-registry/ipi/conf/aws/ipi-conf-aws-commands.sh:54]: you're right. I just fixed it:

> AWS just change my use case from `sa-east-1`, changing the region to `af-south-1` will get the expected results

``` bash
$ yq .controlPlane.platform.aws install-config.yaml
{
  "zones": [
    "af-south-1a",
    "af-south-1b"
  ]
}

$ export LEASED_RESOURCE=af-south-1
$ bash  ipi-conf-aws-commands.sh
zones already set in install-config.yaml, skipped

$ yq .controlPlane.platform.aws install-config.yaml
{
  "zones": [
    "af-south-1a",
    "af-south-1b"
  ],
  "type": "m5.xlarge"
}

$ export LEASED_RESOURCE=us-east-1
$ cp install-config.yaml-use1 install-config.yaml
$ yq .controlPlane.platform.aws install-config.yaml
{
  "zones": [
    "us-east-1a",
    "us-east-1b"
  ]
}

$ bash  ipi-conf-aws-commands.sh
zones already set in install-config.yaml, skipped

$ yq .controlPlane.platform.aws install-config.yaml
{
  "zones": [
    "us-east-1a",
    "us-east-1b"
  ],
  "type": "m6i.xlarge"
}

```

File: ci-operator/step-registry/ipi/conf/aws/ipi-conf-aws-commands.sh (hunk 1 above)
Comment (mtulio) [reply ci-operator/step-registry/ipi/conf/aws/ipi-conf-aws-commands.sh:82]: updating the latest availability:
```
                count(m5.xlarge)  count(m6i.xlarge)  diff(m5_m6i)
region                                                           
af-south-1                     3                  0             3
ap-east-1                      3                  0             3
ap-northeast-1                 3                  2             1
ap-northeast-2                 4                  3             1
ap-northeast-3                 3                  0             3
ap-south-1                     3                  3             0
ap-southeast-1                 3                  3             0
ap-southeast-2                 3                  2             1
ca-central-1                   3                  0             3
eu-central-1                   3                  3             0
eu-north-1                     3                  0             3
eu-south-1                     3                  0             3
eu-west-1                      3                  3             0
eu-west-2                      3                  0             3
eu-west-3                      3                  3             0
me-south-1                     3                  0             3
sa-east-1                      3                  2             1
us-east-1                      5                  5             0
us-east-2                      3                  3             0
us-west-1                      2                  2             0
us-west-2                      6                  4             2

```

**Comment by mtulio** [comment]
@staebler ptal?

File: ci-operator/step-registry/ipi/conf/aws/ipi-conf-aws-commands.sh [hunk 4]
```diff
@@ -5,58 +5,133 @@ set -o errexit
...
+      --region "${REGION}" \
+      --location-type availability-zone \
+      --filters Name=instance-type,Values=\"${type_backup}\" \
+      --query 'InstanceTypeOfferings[].Location' \
+      --output json \
+      | jq -r ".[] |select(.? | match(\"${zones_expr}\"))" \
+      | wc -l)
+  fi
+
+  if [[ ${zones_cnt_pref} -ge ${zones_cnt_backup} ]]; then
```
Comment (staebler) [inline ci-operator/step-registry/ipi/conf/aws/ipi-conf-aws-commands.sh:82]: When you switch this logic to check if there are enough zones for the preferred or backup type, then make sure that the zones with availability matches exactly when the zones are explicitly specified in the install config.

**Comment by mtulio** [comment]
I need to raise the priority on this. Patric, I will work on this next week. Please let me know if you also can see something else to change or any other suggestion.
/cc @patrickdillon 
/reopen

**Comment by droslean** [comment]
@mtulio What is the status of this PR?

**Comment by mtulio** [comment]
> Sure Marco, but I think I need to get caught up. I don't understand the goal here. It looks like you are developing logic to choose the best instance type but what good does that do to put it in ci? Should this logic go in the installer?

@patrickdillon :
This is a follow-up of PR https://github.com/openshift/installer/pull/5327, I am also not sure if there is a good place to implement that logic or if we just need to change the `arch_instance_type=m5` to `arch_instance_type=m6i`.

The goal in this PR is to simulate the installer behavior when the m6i is not available in all zones within the region - but tbh it seems to be very specific and, as you can see, is becoming very complex to be implemented here.

**Comment by mtulio** [comment]
> @mtulio What is the status of this PR?

@droslean sorry I was out since your last message. Let's keep it closed if Patrick agrees with it.
